backend/pkg/commons/db2/parser.go

```diff
@@ -0,0 +1,195 @@
+		return nil, err
+	}
+
+	// When the block is not found, the API returns JSON null.
+	if head == nil {
+		return nil, ethereum.NotFound
+	}
+
+	var body rpcBlock
+	if err := json.Unmarshal(raw, &body); err != nil {
+		return nil, err
+	}
+	// Quick-verify transaction and uncle lists. This mostly helps with debugging the server.
+	if head.UncleHash == types.EmptyUncleHash && len(body.UncleHashes) > 0 {
+		return nil, errors.New("server returned non-empty uncle list but block header indicates no uncles")
+	}
+	if head.UncleHash != types.EmptyUncleHash && len(body.UncleHashes) == 0 {
+		return nil, errors.New("server returned empty uncle list but block header indicates uncles")
+	}
+	if head.TxHash == types.EmptyTxsHash && len(body.Transactions) > 0 {
+		return nil, errors.New("server returned non-empty transaction list but block header indicates no transactions")
+	}
+	if head.TxHash != types.EmptyTxsHash && len(body.Transactions) == 0 {
+		return nil, errors.New("server returned empty transaction list but block header indicates transactions")
+	}
+	// Load uncles because they are not included in the block response.
+	uncles := make([]*types.Header, len(body.UncleHashes))
+	for i := 0; i < len(body.UncleHashes); i++ {
+		err := json.Unmarshal(rawUncles[i].Result, &uncles[i])
+		if err != nil {
+			return nil, err
+		}
+	}
+	// Fill the sender cache of transactions in the block.
+	txs := make([]*types.Transaction, len(body.Transactions))
+	for i, tx := range body.Transactions {
+		if tx.From != nil {
+			setSenderFromDBSigner(tx.tx, *tx.From, body.Hash)
+		}
+		txs[i] = tx.tx
+	}
+	return types.NewBlockWithHeader(head).WithBody(
+		types.Body{
+			Transactions: txs,
+			Uncles:       uncles,
+			Withdrawals:  body.Withdrawals,
+			Requests:     body.Requests,
+		}), nil
+}
+
+// SenderFromDBSigner is a types.Signer that remembers the sender address returned by the RPC
+// server. It is stored in the transaction's sender address cache to avoid an additional
+// request in TransactionSender.
+// copy of senderFromServer
+// https://github.com/ethereum/go-ethereum/blob/v1.14.11/ethclient/signer.go#L30
+type SenderFromDBSigner struct {
+	addr      common.Address
+	Blockhash common.Hash
+}
+
+var errNotCached = errors.New("sender not cached")
+
+func setSenderFromDBSigner(tx *types.Transaction, addr common.Address, block common.Hash) {
+	// Use types.Sender for side-effect to store our signer into the cache.
+	_, _ = types.Sender(&SenderFromDBSigner{addr, block}, tx)
+}
+
+func (s *SenderFromDBSigner) Equal(other types.Signer) bool {
+	os, ok := other.(*SenderFromDBSigner)
+	return ok && os.Blockhash == s.Blockhash
+}
+
+func (s *SenderFromDBSigner) Sender(tx *types.Transaction) (common.Address, error) {
+	if s.addr == (common.Address{}) {
+		return common.Address{}, errNotCached
+	}
+	return s.addr, nil
+}
+
+func (s *SenderFromDBSigner) ChainID() *big.Int {
+	panic("can't sign with SenderFromDBSigner")
```

Comment:
```suggestion
	panic("can't sign with SenderFromDBSigner: ChainID method not supported")
```

Reply:
I prefer to modify as little as possible when we "copy" code from somewhere.
The reasoning behind that is: let say in the future the source is update, we propagate this update, but we loose the local change.
So we may introduce a behavior that is prone to change.